lib/src/stream/mod.rs

```diff
@@ -118,15 +120,47 @@ pub fn encode_stream(
         }
         for worker_thread in worker_threads.iter_mut() {
             let mut buf = vec![];
-            if let Ok(size) = read.by_ref().take(BLOCK_SIZE as u64).read_to_end(&mut buf) {
-                if size == 0 {
-                    finalize = true;
+            let mut rle_data = vec![];
+            let mut rle_total_count = 0;
+            let mut rle_count = 0;
+            let mut rle_last_char = None;
+            while rle_total_count < RLE_LIMIT {
+                // RLE can blow up 4chars to 5, hence we keep a safety margin
+                let to_take = (RLE_LIMIT - rle_data.len()) * 4 / 5;
+                let mut buf_current = vec![];
+                if let Ok(size) = read
+                    .by_ref()
+                    .take(to_take as u64)
+                    .read_to_end(&mut buf_current)
+                {
+                    if size == 0 {
+                        finalize = true;
+                        break;
+                    }
+                } else {
                     break;
                 }
-            } else {
+                let rle_result = rle(&buf_current, rle_count, rle_last_char);
+                let (mut rle_next, rle_next_count, rle_next_char) = rle_result;
+
+                let next_data_len = rle_data.len() + rle_next.len();
+                rle_total_count = rle_total_size(next_data_len, rle_next_count, rle_next_char);
+
+                // Avoid exceeding the limit.
+                if rle_total_count <= RLE_LIMIT {
```

Comment:
I have noticed that this condition is always true, since `to_take` would be equal to 0 when the limit is exceeded, therefore `size` will also be 0 and the loop will be broken after that. I'm going to remove that condition in my next commit.